### PR TITLE
Add or update language with the new packs

### PR DESCRIPTION
**Pull request by maximebiloe**
| Questions | Answers |
| --- | --- |
| Branch? | develop |
| Description? | Add or update language with the new packs and add Basque, Spanish (Peru), Spanish (Venezuelan) and Tamil. |
| Type? | bug fix |
| Category? | CO |
| BC breaks? | no |
| Deprecations? | no |
| Fixed ticket? | http://forge.prestashop.com/browse/BOOM-1602 |
| How to test? | Try Add/Update language in International > Translations and don't forget to test the new languages. |

### CODE REVIEW

File: classes/Language.php [hunk 1]
```diff
@@ -650,6 +616,26 @@ public static function getIsoById($id_lang)
...
+    public static function getJsonLanguageDetails($locale)
+    {
+        if (self::$_cache_all_language_json === null) {
+            self::$_cache_all_language_json = array();
+            $allLanguages = file_get_contents(_PS_ROOT_DIR_.self::ALL_LANGUAGES_FILE);
+            $allLanguages = json_decode($allLanguages, true);
+
+            $jsonLastErrorCode = json_last_error();
+            if (JSON_ERROR_NONE !== $jsonLastErrorCode) {
+                throw new \Exception('The legacy to standard locales JSON could not be decoded', $jsonLastErrorCode);
```
Comment (mickaelandrieu) [inline classes/Language.php]: I'm not sure this is what you want to do here. You should keep `500` as Exception error code and add jsonLastErrorCode in the Exception message.

File: classes/Language.php [hunk 2]
```diff
@@ -903,6 +889,19 @@ public static function isInstalled($iso_code)
         return isset(self::$_cache_language_installation[$iso_code]) ? self::$_cache_language_installation[$iso_code] : false;
     }
 
+    public static function isInstalledByLocale($locale)
```
Comment (mickaelandrieu) [inline classes/Language.php]: PHPDoc could be fine, for ``@return false|id_lang``

File: controllers/admin/AdminTranslationsController.php [hunk 3]
```diff
@@ -37,7 +37,7 @@ class AdminTranslationsControllerCore extends AdminController
     const CONTENT_TYPE_ACCEPTED = array('txt', 'tpl', 'html');
 
     /** @var string : Link which list all pack of language */
-    protected $link_lang_pack = 'http://www.prestashop.com/download/lang_packs/get_each_language_pack.php';
+    protected $link_lang_pack = 'http://i18n.prestashop.com/translations/'._PS_VERSION_.'/available_languages.json';
```
Comment (mickaelandrieu) [inline controllers/admin/AdminTranslationsController.php]: how about a constant that contains ``http://i18n.prestashop.com/`` ?

File: controllers/admin/AdminTranslationsController.php [hunk 4]
```diff
@@ -865,76 +866,26 @@ public static function filesListToPaths($list)
...
-                        foreach ($checks as $check) {
-                            $this->errors[] = sprintf($this->trans('Please check rights for folder and files in %s', array(), 'Admin.Notifications.Error'), $check);
-                        }
-                        if (!unlink($file)) {
-                            $this->errors[] = sprintf($this->trans('Cannot delete the archive %s.', array(), 'Admin.International.Notification'), $file);
-                        }
-                    }
-                } else {
-                    $this->errors[] = $this->trans('The server does not have permissions for writing.', array(), 'Admin.Notifications.Error').' '.sprintf($this->trans('Please check rights for %s', array(), 'Admin.Notifications.Error'), dirname($file));
+                $this->redirect(false, '15');
```
Comment (mickaelandrieu) [inline controllers/admin/AdminTranslationsController.php]: 15 ?

**Review by mickaelandrieu — COMMENTED** [review]
Some questions

File: controllers/admin/AdminTranslationsController.php [hunk 5]
```diff
@@ -666,7 +666,7 @@ public static function addNewTabs($iso_code, $files)
                             }
 
                             if (!Validate::isGenericName($tab->name[(int)$id_lang])) {
-                                $errors[] = sprintf($this->trans('Tab "%s" is not valid', array(), 'Admin.International.Notification'), $tab->name[(int)$id_lang]);
+                                $errors[] = Context::getContext()->getTranslator()->trans('Tab "%s" is not valid', array($tab->name[(int)$id_lang]), 'Admin.International.Notification');
```
Comment (aleeks) [inline controllers/admin/AdminTranslationsController.php]: Why not $this->trans() ?!

File: controllers/admin/AdminTranslationsController.php (hunk 3 above)
Comment (aleeks) [inline controllers/admin/AdminTranslationsController.php]: Agree with @mickaelandrieu

File: controllers/admin/AdminTranslationsController.php [hunk 6]
```diff
@@ -181,18 +181,19 @@ public function initToolbar()
...
-        if ($lang_packs = Tools::file_get_contents($file_name, false, $array_stream_context)) {
-            if ($lang_packs != '' && $lang_packs = json_decode($lang_packs, true)) {
-                foreach ($lang_packs as $key => $lang_pack) {
-                    if (!Language::isInstalled($lang_pack['iso_code'])) {
-                        $packs_to_install[$key] = $lang_pack;
+        $arrayStreamContext = @stream_context_create(array('http' => array('method' => 'GET', 'timeout' => 8)));
+        
+        if ($langPacks = Tools::file_get_contents($this->link_lang_pack, false, $arrayStreamContext)) {
+            if ($langPacks != '' && $langPacks = json_decode($langPacks, true)) {
+                foreach ($langPacks as $locale => $langName) {
```
Comment (aleeks) [inline controllers/admin/AdminTranslationsController.php]: Is it always array `$lang_packs` ?

Comment (maximebiloe) [inline controllers/admin/AdminTranslationsController.php]: Yes, with the second parameters of `json_decode` set to `true` it's an associative array.
http://fr.php.net/manual/en/function.json-decode.php

File: controllers/admin/AdminTranslationsController.php (hunk 5 above)
Comment (maximebiloe) [inline controllers/admin/AdminTranslationsController.php]: There is no `$this` in a static context

Comment (aleeks) [inline controllers/admin/AdminTranslationsController.php]: Oh yes, i didn't look this function declaration :confounded:

File: controllers/admin/AdminTranslationsController.php (hunk 3 above)
Comment (maximebiloe) [inline controllers/admin/AdminTranslationsController.php]: Sorry, I can't.
`self::TRANSLATIONS_HOST.'/translations/%version%/%locale%/%locale%.zip'` won't pass in PHP 5.4

**Comment by aleeks** [comment]
Merged because any new trans! Thank you @maximebiloe

**Pull request merged**